src/eh1/delay.rs

```diff
@@ -1,23 +1,403 @@
-//! Delay mock implementations.
+//! Delay mock implementations, implementing both sync and async
+//! [`DelayNs`](https://docs.rs/embedded-hal/latest/embedded_hal/delay/trait.DelayNs.html)
+//! traits.
+//!
+//! ## Choosing a Delay Implementation
+//!
+//! There are three implementations available depending on your use case:
+//!
+//! - If you want **no actual delay**, create a
+//!   [`NoopDelay`](struct.NoopDelay.html) stub. It will always return
+//!   immediately, without a delay. This is useful for fast tests, where you
+//!   don't actually need to wait for the hardware.
+//! - If you do want the **real delay behavior** when running your tests, use
+//!   [`StdSleep`](struct.StdSleep.html) stub implementation, which uses
+//!   [`std::thread::sleep`](https://doc.rust-lang.org/std/thread/fn.sleep.html)
+//!   to implement the delay.
+//! - For a **configurable delay** implementation that supports expectations,
+//!   use the [`CheckedDelay`](type.CheckedDelay.html) mock. By default it
+//!   doesn't perform an actual delay, but allows the user to enable them
+//!   individually for each expected call.
 //!
 //! ## Usage
 //!
-//! If the actual sleep duration is not important, simply create a
-//! [`NoopDelay`](struct.NoopDelay.html) instance. There will be no actual
-//! delay. This is useful for fast tests, where you don't actually need to wait
-//! for the hardware.
+//! ```
+//! # use eh1 as embedded_hal;
+//! use std::time::Duration;
+//!
+//! use embedded_hal::delay::DelayNs;
+//! use embedded_hal_mock::eh1::delay::{CheckedDelay, NoopDelay, StdSleep, Transaction};
+//!
+//! // No actual delay
+//!
+//! let mut delay = NoopDelay::new();
+//! delay.delay_ms(50); // Returns immediately
+//!
+//! // Real delay
+//!
+//! let mut delay = StdSleep::new();
+//! delay.delay_ms(50); // Will sleep for 50 ms
+//!
+//! // Configurable mock
 //!
-//! If you do want the real delay behavior, use
-//! [`StdSleep`](struct.StdSleep.html) which uses
-//! [`std::thread::sleep`](https://doc.rust-lang.org/std/thread/fn.sleep.html)
-//! to implement the delay.
+//! let transactions = vec![
+//!     Transaction::delay_ns(50_000_000),
+//!     Transaction::delay_us(60_000),
+//!     Transaction::delay_ms(70).wait(),
+//! ];
+//!
+//! let mut delay = CheckedDelay::new(&transactions);
+//!
+//! delay.delay_ms(50); // Conversion to nanoseconds
+//! delay.delay_ms(60); // Conversion to microseconds
+//! delay.delay_ms(70); // This will actually delay
```

Comment:
I had to read the example a few times to understand it. If I understand it correctly, the transactions offer different constructor methods, but in the end the trait only tests for `delay_ns` calls, so they are all interchangeable.

The `delay.delay_ms(...)` calls would be done by the driver-under-test.

This makes sense to test that the implementation of the mock works. However, with the comments you added ("Conversion to nanoseconds / microseconds") it looks as if we would assert that the driver actually calls the `delay_us` and `delay_ns` methods... However, that's not the case.

Maybe my confusion also comes from the fact that the transaction types differentiate by ms/us/ns.

Can we have simpler transactions, and remove `DelayMs` and `DelayUs` in favor of `DelayNs`? We can do the nanosecond calculations directly in the constructor methods (i.e. `pub fn delay_ms(...)` would construct a `DelayNs` transaction).

Reply:
Ah, I see! Yes, I think your suggestion with the big `DelayNs` value makes sense.

Reply:
Ah, I see! Yes, I think your suggestion with the big `DelayNs` value makes sense.